lib/matplotlib/font_manager.py

```diff
@@ -164,19 +160,20 @@ def get_fontext_synonyms(fontext):
 def list_fonts(directory, extensions):
     """
     Return a list of all fonts matching any of the extensions,
-    possibly upper-cased, found recursively under the directory.
+    found recursively under the directory.
     """
-    pattern = ';'.join(['*.%s;*.%s' % (ext, ext.upper())
-                        for ext in extensions])
-    return cbook.listFiles(directory, pattern)
+    return [str(path)
+            for ext in extensions
+            for path in filter(Path.is_file,
+                               Path(directory).rglob("*.{}".format(ext)))]
```

Comment:
This walks the directory tree as many times as there are extensions instead of just once.

Reply:
fixed